sdk/core/Azure.Core/src/Serialization/ModelSerializer.cs

```diff
@@ -0,0 +1,183 @@
+        /// <returns>A <paramref name="returnType"/> representation of the <see cref="BinaryData"/>.</returns>
+        /// <exception cref="InvalidOperationException">Throws if <paramref name="returnType"/> does not implement <see cref="IModelSerializable{T}"/>.</exception>
+        /// <exception cref="InvalidOperationException">Throws if <paramref name="returnType"/> does not have a public or internal parameterless constructor.</exception>
+        /// <exception cref="FormatException">If the model does not support the requested <see cref="ModelSerializerOptions.Format"/>.</exception>
+        public static object Deserialize(BinaryData data, [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors | DynamicallyAccessedMemberTypes.NonPublicConstructors)] Type returnType, ModelSerializerOptions? options = default)
+        {
+            options ??= ModelSerializerOptions.DefaultWireOptions;
+
+            return GetInstance(returnType).Deserialize(data, options);
+        }
+
+        /// <summary>
+        /// Converts the <see cref="BinaryData"/> into a <paramref name="returnType"/>.
+        /// </summary>
+        /// <param name="data">The <see cref="BinaryData"/> to convert.</param>
+        /// <param name="returnType">The type of the objec to convert and return.</param>
+        /// <param name="format">The <see cref="ModelSerializerFormat"/> to use.</param>
+        /// <returns>A <paramref name="returnType"/> representation of the <see cref="BinaryData"/>.</returns>
+        /// <exception cref="InvalidOperationException">Throws if <paramref name="returnType"/> does not implement <see cref="IModelSerializable{T}"/>.</exception>
+        /// <exception cref="InvalidOperationException">Throws if <paramref name="returnType"/> does not have a public or internal parameterless constructor.</exception>
+        /// <exception cref="FormatException">If the model does not support the requested <see cref="ModelSerializerFormat"/>.</exception>
+        public static object Deserialize(BinaryData data, [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors | DynamicallyAccessedMemberTypes.NonPublicConstructors)] Type returnType, ModelSerializerFormat format)
+            => Deserialize(data, returnType, ModelSerializerOptions.GetOptions(format));
+
+        /// <summary>
+        /// Converts the value of a model into a <see cref="BinaryData"/>.
+        /// </summary>
+        /// <param name="model">The model to convert.</param>
+        /// <param name="options">The <see cref="ModelSerializerOptions"/> to use.</param>
+        /// <returns>A <see cref="BinaryData"/> representation of the model in the <see cref="ModelSerializerFormat"/> specified by the <paramref name="options"/>.</returns>
+        public static BinaryData SerializeCore(IModelJsonSerializable<object> model, ModelSerializerOptions options)
+        {
+            using ModelWriter writer = new ModelWriter(model, options);
```

Comment:
I'm sure I'm missing something, but how does the buffer pooling help if we are creating a new ModelWriter instance, and each ModelWriter has its own SequenceBuilder instance?

Reply:
Ah I guess the memory pooling is per model Serialize call.